Suggest use of `--print target-list` when target is not found.

tmiasko opened the pull request:
If given target could not be found suggest using `--print target-list`.
Previously, error has been reported as:

$ rustc --target 86-unknown-linux-gnu
error: Error loading target specification: Could not find specification for target "86-unknown-linux-gnu"

After changes it looks as follows:

$ rustc --target 86-unknown-linux-gnu
error: Error loading target specification: Could not find specification for target "x86-unknown-linux-gnu". Use `--print target-list` for a list of supported targets

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @arielb1 (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

nagisa commented:
This needs a more careful wording. --print target-list does not print *all* targets that would be accepted in any given environment. There’s custom json targets, `RUST_TARGET_PATH` and similar stuff, all of which are not covered by `--print target-list`.

What this should be saying instead is:

```
error: Error loading target specification: Could not find specification for target "x86-unknown-linux-gnu". Use `--print target-list` for a list of built-in targets.
```

Better off you could make the note/help an actual note:

```
error: Error loading target specification: Could not find specification for target "x86-unknown-linux-gnu". 
help: Use `--print target-list` for a list of built-in targets.
```

For that you’ll need to convert this code to use the DiagnosticBuilder. Namely the code would end up looking something like

```
panic!(sp.struct_fatal("Error loading…").help("Use `--print target-list`…").emit());
```

tmiasko commented:
Thanks @nagisa. I updated the message and moved it to help sub-section. Though, I had to manually panic with `errors::FatalError`, otherwise main driver does not understand that panic is not unexpected (this happens because `emit()` doesn't return anything).

nagisa commented:
r=me

eddyb commented:
@bors r=nagisa

bors commented:
:pushpin: Commit 443f1ca has been approved by `nagisa`

<!-- @bors r=nagisa 443f1ca83c8b323d530ba62ed1974a75f94e0e17 -->

bors commented:
:hourglass: Testing commit 443f1ca with merge 5556554...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/10057), [auto-linux-32-opt](https://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/10104), [auto-linux-64-cargotest](https://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/1347), [auto-linux-64-cross-freebsd](https://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/1173), [auto-linux-64-debug-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/3324), [auto-linux-64-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/10042), [auto-linux-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/10013), [auto-linux-64-opt-mir](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-mir/builds/1608), [auto-linux-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/1986), [auto-linux-64-x-android-t](https://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/9990), [auto-linux-cross-opt](https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/3214), [auto-linux-musl-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/5164), [auto-mac-32-opt](https://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/10080), [auto-mac-64-opt](https://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/10060), [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/2007), [auto-mac-cross-ios-opt](https://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/1183), [auto-win-gnu-32-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/5164), [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/2035), [auto-win-gnu-64-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/5149), [auto-win-msvc-32-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/4338), [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/1365), [auto-win-msvc-64-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/5149), [auto-win-msvc-64-opt-mir](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-mir/builds/1643), [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/2029)
Approved by: nagisa
Pushing 5556554e2504ba2df8712ad0666ce1a11915dade to master...
